File: IFATCCompanion/Weather/RadarPrecipitationProvider.swift
import Foundation
import CoreLocation
import CoreGraphics
import MapKit

// MARK: - Layer type
//
// Precipitation overlays come from two fundamentally different kinds of source
// and MUST be labeled accordingly. NOAA and EUMETNET OPERA are **true radar**;
// NASA GPM IMERG (via GIBS) is a **satellite precipitation estimate** and is never
// presented as radar. The app never shows the phrase "global radar".

enum PrecipitationLayerType {
    case radar             // NOAA/NWS, EUMETNET OPERA
    case satelliteEstimate // NASA GPM IMERG / GIBS

    /// The user-facing label for this layer type.
    var uiLabel: String {
        switch self {
        case .radar: return "Radar precipitation"
        case .satelliteEstimate: return "Satellite precipitation estimate"
        }
    }

    var isRadar: Bool { self == .radar }
}

// MARK: - Provider protocol
//
// A precipitation overlay provider for a map region. All fetches are keyless and
// free, and every conformer must be compatible with commercial app inclusion,
// redistribution/display, and attribution-only terms. Providers that would require
// a paid plan, user account, API-key billing, or non-commercial-only terms are out
// of scope. Selection order is NOAA → OPERA → NASA (see `PrecipitationOverlayService`).

protocol RadarPrecipitationProvider {
    var id: String { get }
    var displayName: String { get }
    var coverageDescription: String { get }
    var attributionText: String? { get }
    /// Whether this provider serves *true observed radar* (vs a satellite estimate
    /// or a mock stand-in). Never advertise satellite/mock data as true radar.
    var supportsTrueRadar: Bool { get }
    /// Radar vs satellite-estimate — drives the UI label and confidence.
    var layerType: PrecipitationLayerType { get }
    /// Coarse confidence in this layer (radar high, satellite lower).
    var confidence: HazardConfidence { get }

    /// Whether the provider covers a region (synchronous coverage check — no I/O).
    func covers(region: MKCoordinateRegion) -> Bool
    /// Whether the provider can actually **render an overlay** for the region right
    /// now — a stricter check than `covers`. A provider may geographically cover a
    /// region yet be unable to produce imagery there (e.g. no data endpoint is
    /// configured), in which case it must not be selected as the active overlay and
    /// selection falls through to the next provider. Defaults to `covers` (a provider
    /// that covers a region can render it) — override to gate on a live capability.
    func canRenderOverlay(for region: MKCoordinateRegion) -> Bool
    /// Whether the provider covers the region (async form; defaults to `covers`).
    func isAvailable(for region: MKCoordinateRegion) async -> Bool
    /// The time steps available for the region.
    func availableFrames(for region: MKCoordinateRegion) async throws -> [RadarFrame]
    /// XYZ tile URL for a frame, when the provider is tiled (else nil).
    func overlayTileURL(z: Int, x: Int, y: Int, frame: RadarFrame) async throws -> URL?
    /// Rendered PNG for a bounding box (fetches bytes).
    func exportImage(for bbox: RadarBoundingBox, size: CGSize, frame: RadarFrame) async throws -> Data?
    /// A synchronous URL for the rendered image, for SwiftUI `AsyncImage`. Nil when
    /// the provider cannot render an image for the region (fail gracefully).
    func exportImageURL(for bbox: RadarBoundingBox, size: CGSize, frame: RadarFrame?) -> URL?
}

extension RadarPrecipitationProvider {
    /// Default: a provider that geographically covers a region can also render it.
    /// Override where rendering depends on a live capability (e.g. a configured or
    /// reachable data source) that coverage alone doesn't guarantee.
    func canRenderOverlay(for region: MKCoordinateRegion) -> Bool { covers(region: region) }
    /// Default async availability delegates to the synchronous coverage check.
    func isAvailable(for region: MKCoordinateRegion) async -> Bool { covers(region: region) }
    /// Default: no synchronous image URL (conformers override where they can render).
    func exportImageURL(for bbox: RadarBoundingBox, size: CGSize, frame: RadarFrame?) -> URL? { nil }
    /// The user-facing layer label (radar vs satellite estimate).
    var uiLayerLabel: String { layerType.uiLabel }
}

// MARK: - NOAA / NWS

/// NOAA/NWS radar base reflectivity (MRMS) overlay provider. Public, keyless NWS
/// ArcGIS radar ImageServer. Coverage is limited to NOAA-covered regions
/// (contiguous U.S. + Alaska/Hawaii/Puerto Rico approximations).
struct NOAARadarPrecipitationProvider: RadarPrecipitationProvider {

    var baseURL: String

    init(baseURL: String = "https://mapservices.weather.noaa.gov/eventdriven/rest/services/radar/radar_base_reflectivity_time/ImageServer") {
        self.baseURL = baseURL
    }

    let id = "noaa-nws-radar"
    let displayName = "NOAA/NWS radar precipitation"
    let coverageDescription = "Available in NOAA-covered radar regions"
    let attributionText: String? = "Radar precipitation data: NOAA/NWS"
    let supportsTrueRadar = true
    let layerType: PrecipitationLayerType = .radar
    let confidence: HazardConfidence = .high

    /// Approximate NOAA radar coverage boxes (conservative; NOAA-covered regions
    /// only — never implying global coverage).
    static let coverageBoxes: [RadarBoundingBox] = [
        RadarBoundingBox(minLatitude: 20, minLongitude: -130, maxLatitude: 52, maxLongitude: -60),  // CONUS
        RadarBoundingBox(minLatitude: 50, minLongitude: -180, maxLatitude: 72, maxLongitude: -129), // Alaska
        RadarBoundingBox(minLatitude: 18, minLongitude: -161, maxLatitude: 23, maxLongitude: -154), // Hawaii
        RadarBoundingBox(minLatitude: 16, minLongitude: -68, maxLatitude: 20, maxLongitude: -64)     // Puerto Rico
    ]

    func covers(region: MKCoordinateRegion) -> Bool { Self.covers(region: region) }

    static func covers(region: MKCoordinateRegion) -> Bool {
        let box = RadarBoundingBox(region: region)
        return coverageBoxes.contains { $0.overlaps(box) }
    }

    static func covers(coordinate: CLLocationCoordinate2D) -> Bool {
        coverageBoxes.contains { $0.contains(coordinate) }
    }

    func availableFrames(for region: MKCoordinateRegion) async throws -> [RadarFrame] {
        guard covers(region: region) else { return [] }
        return [RadarFrame(id: "current", timestamp: Date(), isForecast: false, label: "Current")]
    }

    func overlayTileURL(z: Int, x: Int, y: Int, frame: RadarFrame) async throws -> URL? { nil }

    func exportImage(for bbox: RadarBoundingBox, size: CGSize, frame: RadarFrame) async throws -> Data? {
        guard let url = exportImageURL(for: bbox, size: size, frame: frame) else { return nil }
        var request = URLRequest(url: url)
        request.timeoutInterval = 12
        request.setValue("IFATCCompanion/1.0", forHTTPHeaderField: "User-Agent")
        let (data, response) = try await URLSession.shared.data(for: request)
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) { return nil }
        return data.isEmpty ? nil : data
    }

    func exportImageURL(for bbox: RadarBoundingBox, size: CGSize, frame: RadarFrame?) -> URL? {
        guard size.width > 0, size.height > 0 else { return nil }
        var components = URLComponents(string: "\(baseURL)/exportImage")
        // 4326 bbox, rendered in Web Mercator (3857) to align with MapKit.
        let bboxValue = "\(bbox.minLongitude),\(bbox.minLatitude),\(bbox.maxLongitude),\(bbox.maxLatitude)"
        components?.queryItems = [
            URLQueryItem(name: "bbox", value: bboxValue),
            URLQueryItem(name: "bboxSR", value: "4326"),
            URLQueryItem(name: "imageSR", value: "3857"),
            URLQueryItem(name: "size", value: "\(Int(size.width.rounded())),\(Int(size.height.rounded()))"),
            URLQueryItem(name: "format", value: "png"),
            URLQueryItem(name: "transparent", value: "true"),
            URLQueryItem(name: "f", value: "image")
        ]
        return components?.url
    }
}

// MARK: - Mock

/// A deterministic, offline radar stand-in for Mock Mode and tests. Advertises
/// itself as NOT true radar and serves precipitation as vector cells (drawn as
/// polygons), never as an image claiming to be observed radar.
struct MockRadarPrecipitationProvider: RadarPrecipitationProvider {
    let id = "mock-radar"
    let displayName = "Simulated radar (Mock Mode)"
    let coverageDescription = "Simulated coverage for the mock flight"
    let attributionText: String? = "Simulated precipitation — Mock Mode"
    let supportsTrueRadar = false
    let layerType: PrecipitationLayerType = .radar
    let confidence: HazardConfidence = .high

    func covers(region: MKCoordinateRegion) -> Bool { true }
    func availableFrames(for region: MKCoordinateRegion) async throws -> [RadarFrame] {
        [RadarFrame(id: "mock-current", timestamp: Date(), isForecast: false, label: "Current (mock)")]
    }
    func overlayTileURL(z: Int, x: Int, y: Int, frame: RadarFrame) async throws -> URL? { nil }
    func exportImage(for bbox: RadarBoundingBox, size: CGSize, frame: RadarFrame) async throws -> Data? { nil }
}

File: Android/core/src/main/kotlin/com/h3consultingpartners/ifatccompanion/core/weather/radar/RadarPrecipitationProvider.kt
package com.h3consultingpartners.ifatccompanion.core.weather.radar

import com.h3consultingpartners.ifatccompanion.core.config.AppConfig
import com.h3consultingpartners.ifatccompanion.core.geo.Coordinate
import com.h3consultingpartners.ifatccompanion.core.net.AppHttp
import com.h3consultingpartners.ifatccompanion.core.net.HttpFetching
import com.h3consultingpartners.ifatccompanion.core.net.HttpResult
import com.h3consultingpartners.ifatccompanion.core.platform.Clock
import com.h3consultingpartners.ifatccompanion.core.weather.deviation.HazardConfidence
import com.h3consultingpartners.ifatccompanion.core.weather.deviation.RadarBoundingBox

/**
 * A precipitation overlay provider for a map region.
 *
 * All fetches are keyless and free, and every implementation must be compatible with
 * commercial app inclusion, redistribution/display, and attribution-only terms.
 * Providers that would require a paid plan, user account, API-key billing, or
 * non-commercial-only terms are out of scope. Selection order is
 * NOAA → OPERA → NASA (see [PrecipitationOverlayService]).
 *
 * Ported from `IFATCCompanion/Weather/RadarPrecipitationProvider.swift`. The Swift
 * protocol's default-implementation extension becomes default interface methods.
 */
interface RadarPrecipitationProvider {
    val id: String
    val displayName: String
    val coverageDescription: String
    val attributionText: String?

    /**
     * Whether this provider serves *true observed radar* (vs a satellite estimate or a
     * mock stand-in). Never advertise satellite/mock data as true radar.
     */
    val supportsTrueRadar: Boolean

    /** Radar vs satellite-estimate — drives the UI label and confidence. */
    val layerType: PrecipitationLayerType

    /** Coarse confidence in this layer (radar high, satellite lower). */
    val confidence: HazardConfidence

    /** Whether the provider covers a region (synchronous coverage check — no I/O). */
    fun covers(region: MapRegion): Boolean

    /**
     * Whether the provider can actually **render an overlay** for the region right now —
     * a stricter check than [covers]. A provider may geographically cover a region yet be
     * unable to produce imagery there (e.g. no data endpoint is configured), in which case
     * it must not be selected as the active overlay and selection falls through to the
     * next provider. Defaults to [covers] — override to gate on a live capability.
     */
    fun canRenderOverlay(region: MapRegion): Boolean = covers(region)

    /** Whether the provider covers the region (suspending form; defaults to [covers]). */
    suspend fun isAvailable(region: MapRegion): Boolean = covers(region)

    /** The time steps available for the region. */
    suspend fun availableFrames(region: MapRegion): List<RadarFrame>

    /** XYZ tile URL for a frame, when the provider is tiled (else null). */
    suspend fun overlayTileUrl(z: Int, x: Int, y: Int, frame: RadarFrame): String? = null

    /** Rendered PNG bytes for a bounding box (fetches bytes). */
    suspend fun exportImage(bbox: RadarBoundingBox, size: PixelSize, frame: RadarFrame): ByteArray?

    /**
     * A synchronous URL for the rendered image, for the map to load directly. Null when
     * the provider cannot render an image for the region (fail gracefully).
     */
    fun exportImageUrl(bbox: RadarBoundingBox, size: PixelSize, frame: RadarFrame?): String? = null

    /** The user-facing layer label (radar vs satellite estimate). */
    val uiLayerLabel: String get() = layerType.uiLabel
}

/**
 * Shared image-fetch behaviour: every provider that renders server-side downloads its
 * PNG the same way — revalidating cache policy, 12-second timeout, and the app's
 * descriptive User-Agent. Empty bodies and non-2xx statuses degrade to null rather than
 * throwing, so a provider outage blanks the layer instead of failing the flight.
 */
internal suspend fun fetchOverlayPng(http: HttpFetching, url: String): ByteArray? {
    val result = http.get(
        url,
        headers = mapOf(
            "User-Agent" to AppHttp.userAgent,
            // The OkHttp equivalent of iOS's `.reloadRevalidatingCacheData`.
            "Cache-Control" to "max-age=0",
        ),
        timeoutSeconds = OVERLAY_IMAGE_TIMEOUT_SECONDS,
    )
    val response = (result as? HttpResult.Success)?.response ?: return null
    if (!response.isSuccess) return null
    return response.body.takeIf { it.isNotEmpty() }
}

internal const val OVERLAY_IMAGE_TIMEOUT_SECONDS = 12L

/**
 * Percent-encode a query parameter value. Every overlay URL is assembled by hand here
 * rather than with a URI builder, because `:core` is plain Kotlin/JVM and the exact
 * parameter order matters for the ArcGIS/WMS endpoints' own caches.
 */
internal fun encodeQuery(value: String): String = buildString {
    for (byte in value.toByteArray(Charsets.UTF_8)) {
        val ch = byte.toInt().toChar()
        when {
            ch.isLetterOrDigit() && ch.code < 128 -> append(ch)
            ch == '-' || ch == '_' || ch == '.' || ch == '~' -> append(ch)
            else -> append('%').append("%02X".format(byte.toInt() and 0xFF))
        }
    }
}

internal fun queryString(items: List<Pair<String, String>>): String =
    items.joinToString("&") { (name, value) -> "$name=${encodeQuery(value)}" }

/**
 * NOAA/NWS radar base reflectivity (MRMS) overlay provider. Public, keyless NWS ArcGIS
 * radar ImageServer. Coverage is limited to NOAA-covered regions (contiguous U.S. plus
 * Alaska/Hawaii/Puerto Rico approximations).
 */
class NOAARadarPrecipitationProvider(
    private val http: HttpFetching,
    private val clock: Clock = Clock.system,
    private val baseUrl: String = AppConfig.Endpoints.NWS_RADAR_IMAGE_SERVER,
) : RadarPrecipitationProvider {

    override val id = "noaa-nws-radar"
    override val displayName = "NOAA/NWS radar precipitation"
    override val coverageDescription = "Available in NOAA-covered radar regions"
    override val attributionText: String? = "Radar precipitation data: NOAA/NWS"
    override val supportsTrueRadar = true
    override val layerType = PrecipitationLayerType.RADAR
    override val confidence = HazardConfidence.HIGH

    override fun covers(region: MapRegion): Boolean = Companion.covers(region)

    override suspend fun availableFrames(region: MapRegion): List<RadarFrame> {
        if (!covers(region)) return emptyList()
        return listOf(
            RadarFrame(id = "current", timestampMillis = clock.nowMillis(), isForecast = false, label = "Current"),
        )
    }

    override suspend fun exportImage(bbox: RadarBoundingBox, size: PixelSize, frame: RadarFrame): ByteArray? {
        val url = exportImageUrl(bbox, size, frame) ?: return null
        return fetchOverlayPng(http, url)
    }

    override fun exportImageUrl(bbox: RadarBoundingBox, size: PixelSize, frame: RadarFrame?): String? {
        if (!size.isValid) return null
        // 4326 bbox, rendered in Web Mercator (3857) to align with the map's projection.
        val bboxValue =
            "${bbox.minLongitude},${bbox.minLatitude},${bbox.maxLongitude},${bbox.maxLatitude}"
        val query = queryString(
            listOf(
                "bbox" to bboxValue,
                "bboxSR" to "4326",
                "imageSR" to "3857",
                "size" to "${size.width},${size.height}",
                "format" to "png",
                "transparent" to "true",
                "f" to "image",
            ),
        )
        return "$baseUrl/exportImage?$query"
    }

    companion object {
        /**
         * Approximate NOAA radar coverage boxes (conservative; NOAA-covered regions only
         * — never implying global coverage).
         */
        val coverageBoxes: List<RadarBoundingBox> = listOf(
            RadarBoundingBox(20.0, -130.0, 52.0, -60.0),    // CONUS
            RadarBoundingBox(50.0, -180.0, 72.0, -129.0),   // Alaska
            RadarBoundingBox(18.0, -161.0, 23.0, -154.0),   // Hawaii
            RadarBoundingBox(16.0, -68.0, 20.0, -64.0),     // Puerto Rico
        )

        fun covers(region: MapRegion): Boolean {
            val box = region.boundingBox
            return coverageBoxes.any { it.overlaps(box) }
        }

        fun covers(coordinate: Coordinate): Boolean = coverageBoxes.any { coordinate in it }
    }
}

/**
 * A deterministic, offline radar stand-in for Mock Mode and tests. Advertises itself as
 * NOT true radar and serves precipitation as vector cells (drawn as polygons), never as
 * an image claiming to be observed radar.
 */
class MockRadarPrecipitationProvider(
    private val clock: Clock = Clock.system,
) : RadarPrecipitationProvider {

    override val id = "mock-radar"
    override val displayName = "Simulated radar (Mock Mode)"
    override val coverageDescription = "Simulated coverage for the mock flight"
    override val attributionText: String? = "Simulated precipitation — Mock Mode"
    override val supportsTrueRadar = false
    override val layerType = PrecipitationLayerType.RADAR
    override val confidence = HazardConfidence.HIGH

    override fun covers(region: MapRegion): Boolean = true

    override suspend fun availableFrames(region: MapRegion): List<RadarFrame> = listOf(
        RadarFrame(
            id = "mock-current",
            timestampMillis = clock.nowMillis(),
            isForecast = false,
            label = "Current (mock)",
        ),
    )

    override suspend fun exportImage(bbox: RadarBoundingBox, size: PixelSize, frame: RadarFrame): ByteArray? = null
}
